common/src/stack/command/stack/commands/report/vm/__init__.py

```
@@ -0,0 +1,243 @@
+
+		# Check if the hypervisor is valid
+		# if the param is set
+		if hypervisor and not self.is_hypervisor(hypervisor):
+			raise ParamError(self, 'hypervisor', '{hypervisor} is not a valid hypervisor')
+
+		for vm in vm_hosts:
+			self.bootorder = 1
+			curr_hypervisor = self.get_hypervisor_by_name(vm)
+			loc = conf_loc
+
+			# If the hypervisor param is set
+			# ignore any VM not belonging to the
+			# specified hypervisor host
+			if hypervisor and curr_hypervisor != hypervisor:
+				continue
+
+			conf_loc_attr = self.getHostAttr(curr_hypervisor, 'vm.config.location')
+
+			if conf_loc_attr:
+				loc = conf_loc_attr
+
+			# Handle if no disks are defined for a
+			# virtual machine
+			if vm not in vm_disks:
+				curr_disks = []
+			else:
+				curr_disks = vm_disks[vm]
+
+			template_vars = {}
+			vm_values = vm_info[vm]
+
+			# Assign libvirt template values
+			template_vars['name'] = vm
+			template_vars['memory'] = int(vm_values['memory']) * 1024
+			template_vars['cpucount'] = vm_values['cpu']
+			template_vars['uuid'] = uuid.uuid4()
+
+			# Get template info that varies between VM's
+			template_vars['interfaces'] = self.gen_interfaces(vm)
+			template_vars['disks'] = self.gen_disks(vm, curr_disks)
+
+			vm_config = libvirt_template.render(template_vars)
+			if not bare_output:
+				config_file = Path(f'{loc}/{vm}.xml')
+
+				# Output SUX
+				if len(vm_hosts) > 1:
+					self.addOutput(vm, f'<stack:file stack:name={config_file}>')
+				else:
+					self.addOutput('', f'<stack:file stack:name={config_file}>')
+				self.addOutput('', vm_config)
+				self.addOutput('', '</stack:file>')
+			else:
+				self.addOutput(vm, vm_config)
+		self.endOutput(padChar='', trimOwner=True)
+
+	def getMAC(self):
+		"""
+		Generate a random mac address for
+		virtual machine interfaces
+		"""
+
+		r = random.SystemRandom()
+		m = "%06x" % r.getrandbits(24)
+
+		return "52:54:00:%s:%s:%s" % (m[0:2], m[2:4], m[4:6])
+
+	def getInterfaceByNetwork(self, host, network):
```

probably nuke this- this assumption does not hold up.

---

Looking over the code again, this function was used for mapping what source interface on from the hypervisor (such as br0, br1...etc) goes to the VM via the network names. Is this assumption then that doesn't hold up? If so what is a better way to pick the interfaces to assign to the VMs?

---

Putting it here for reference, we agreed that the best approach for now (until another use case is needed) is to continue the current behavior of assigning the first interface on the hypervisor with the same network as the VM interface, but raise a CommandError if there are no matches. The previous command behavior just silently skipped the output of an interface if no matching network could be found on the hypervisor.

---

Adding another layer to this, originally I was raising a CommandError for ipmi interfaces as well. I see now that instead those interfaces should be skipped instead as they could be a virtual interface but don't need to be defined in the VM config.